File: FB Home/HangerView.m
//
//  HangerView.m
//  FB Home
//
//  Created by Mox Soini on 31.8.2013.
//  Copyright (c) 2013 Mox Soini. All rights reserved.
//

#import "HangerView.h"

@interface HangerView ()

@property (nonatomic) CGPoint sPt;
@property (nonatomic) CGPoint ePt;
@property (nonatomic) CGFloat lineLength;

@end

@implementation HangerView

- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        // Initialization code
        _sPt = CGPointMake(self.bounds.size.width/2, self.bounds.size.height/2);
        _lineLength = self.bounds.size.height/2;
        _ePt = CGPointMake(_sPt.x, _sPt.y + _lineLength);
        [self setOpaque:NO];
    }
    return self;
}


- (void) setAngle:(double)angle
{
    _ePt = CGPointMake(_sPt.x + _lineLength * sin(angle),
                           _sPt.y + _lineLength * cos(angle));
}


// Only override drawRect: if you perform custom drawing.
// An empty implementation adversely affects performance during animation.
- (void)drawRect:(CGRect)rect
{
    [super drawRect:rect];

    // Drawing code    
    
    CGContextRef context = UIGraphicsGetCurrentContext();
    
    // black filled circle
    CGContextSetFillColorWithColor(context, [UIColor blackColor].CGColor);
    CGContextAddEllipseInRect(context, CGRectMake(_sPt.x - _lineLength, _sPt.y - _lineLength, _lineLength*2, _lineLength*2));
    CGContextFillPath(context);

    // white ticks
    CGContextSetStrokeColorWithColor(context, [UIColor whiteColor].CGColor);
    CGContextSetLineWidth(context, 1.0);
    for (int i = 0; i < 8; i++) {
        CGFloat angle = i * 2 * M_PI / 8;
        CGFloat dx = _lineLength * sin(angle);
        CGFloat dy = _lineLength * cos(angle);
        CGContextMoveToPoint(context, _sPt.x + 5.0/6.0*dx, _sPt.y + 5.0/6.0*dy); //start at this point
        CGContextAddLineToPoint(context, _sPt.x + dx, _sPt.y + dy); //draw to this point
    }
    for (int i = 0; i < 24; i++) {
        CGFloat angle = i * 2 * M_PI / 24;
        CGFloat dx = _lineLength * sin(angle);
        CGFloat dy = _lineLength * cos(angle);
        CGContextMoveToPoint(context, _sPt.x + 11.0/12.0*dx, _sPt.y + 11.0/12.0*dy); //start at this point
        CGContextAddLineToPoint(context, _sPt.x + dx, _sPt.y + dy); //draw to this point
    }
    CGContextStrokePath(context);
    
    //red arrow
    CGContextSetStrokeColorWithColor(context, [UIColor redColor].CGColor);
    CGContextSetLineWidth(context, 2.0);
    CGContextMoveToPoint(context, _sPt.x, _sPt.y); //start at this point
    CGContextAddLineToPoint(context, _ePt.x, _ePt.y); //draw to this point
    CGContextStrokePath(context);
    
}


@end

File: FB Home/HangerView.swift
//
//  HangerView.swift
//  FB Home
//
//  Created by Mox Soini on 6.5.2015.
//  Copyright (c) 2015 Mox Soini. All rights reserved.
//

import UIKit

class HangerView : UIView {

  private lazy var sPt: CGPoint = { return CGPoint(x: self.bounds.size.width/2, y: self.bounds.size.height/2) }()
  private lazy var lineLength: CGFloat = { return self.bounds.size.height/2 }()
  private lazy var ePt: CGPoint = { return CGPoint(x: self.sPt.x, y: self.sPt.y + self.lineLength) }()

  var angle: CGFloat = 0 {
    didSet {
      ePt = CGPoint(x: sPt.x + lineLength * sin(angle), y: sPt.y + lineLength * cos(angle))
    }
  }

  override func didMoveToSuperview() {
    isOpaque = false
  }

  // Only override drawRect: if you perform custom drawing.
  override func draw(_ rect: CGRect) {

    super.draw(rect)

    // Drawing code

    guard let context = UIGraphicsGetCurrentContext() else { return }

    // black filled circle
    context.setFillColor(UIColor.black().cgColor)
    context.addEllipse(inRect: CGRect(x: sPt.x - lineLength, y: sPt.y - lineLength, width: lineLength*2, height: lineLength*2))
    context.fillPath()

    // white ticks
    context.setStrokeColor(UIColor.white().cgColor)
    context.setLineWidth(1.0)
    for i in 0..<8 {
      let angle = CGFloat(i) * 2 * CGFloat(M_PI) / 8
      let dx = lineLength * sin(angle)
      let dy = lineLength * cos(angle)
      context.moveTo(x: sPt.x + 5.0/6.0*dx, y: sPt.y + 5.0/6.0*dy) //start at this point
      context.addLineTo(x: sPt.x + dx, y: sPt.y + dy) //draw to this point
    }
    for i in 0..<24 {
      let angle = CGFloat(i) * 2 * CGFloat(M_PI) / 24;
      let dx = lineLength * sin(angle)
      let dy = lineLength * cos(angle)
      context.moveTo(x: sPt.x + 11.0/12.0*dx, y: sPt.y + 11.0/12.0*dy) //start at this point
      context.addLineTo(x: sPt.x + dx, y: sPt.y + dy) //draw to this point
    }
    context.strokePath()

    //red arrow
    context.setStrokeColor(UIColor.red().cgColor)
    context.setLineWidth(2.0)
    context.moveTo(x: sPt.x, y: sPt.y) //start at this point
    context.addLineTo(x: ePt.x, y: ePt.y) //draw to this point
    context.strokePath()

  }

}